In [1]:
import statistics as stats

data = [10, 12, 14, 16, 18, 20]

print("Mean:", stats.mean(data))
print("Median:", stats.median(data))
print("Mode:", stats.mode(data))
print("Min:", min(data))
print("Max:", max(data))
print("Range:", max(data) - min(data))

Mean: 15
Median: 15.0
Mode: 10
Min: 10
Max: 20
Range: 10


Category	Statistics
-  
Central tendency - Mean, Median, Mode  
Dispersion - Range, Variance, Standard Deviation, IQR  
Position - Percentile, Quartile, Decile  
Shape -	Skewness, Kurtosis  
Relationship - Covariance, Correlation  
Probability / Inference	- Z-score, p-value, confidence interval  
Counts - Frequency, Relative frequency  

In [11]:
import numpy as np
from scipy import stats

# -----------------------------
# 1. Descriptive Statistics
# -----------------------------

data = np.array([10, 12, 14, 16, 18, 20])

print("Mean:", np.mean(data))
print("Median:", np.median(data))

# Population variance and SD
print("Population Variance:", np.var(data))
print("Population SD:", np.std(data))

# Sample variance and SD
print("Sample Variance:", np.var(data, ddof=1))
print("Sample SD:", np.std(data, ddof=1))

print("Q1:", np.percentile(data, 25))
print("Q3:", np.percentile(data, 75))
print("IQR:", stats.iqr(data))

print("Skewness:", stats.skew(data))
print("Kurtosis:", stats.kurtosis(data))

# Z-scores
print("Z-scores - Population:", stats.zscore(data))  # uses ddof=0 → population SD
print("Z-scores - Sample:", stats.zscore(data, ddof=1)) # uses ddof=1 → sample SD


# -----------------------------
# 2. Pearson Correlation
# -----------------------------

x = np.array([1, 2, 3, 4, 5])
y = np.array([2, 4, 5, 8, 10])

correlation = stats.pearsonr(x, y)

print("\nPearson Correlation:")
print("Correlation coefficient:", correlation.statistic)
print("P-value:", correlation.pvalue)


# -----------------------------
# 3. Independent t-test
# -----------------------------

group1 = np.array([10, 12, 14, 15, 13])
group2 = np.array([20, 22, 19, 21, 23])

t_test = stats.ttest_ind(group1, group2)

print("\nIndependent t-test:")
print("T-statistic:", t_test.statistic)
print("P-value:", t_test.pvalue)


# -----------------------------
# 4. Chi-square Test
# -----------------------------

# Observed frequency table
table = np.array([
    [30, 20],
    [10, 40]
])

chi_square = stats.chi2_contingency(table)

print("\nChi-square test:")
print("Chi-square statistic:", chi_square.statistic)
print("P-value:", chi_square.pvalue)
print("Degrees of freedom:", chi_square.dof)
print("Expected frequencies:")
print(chi_square.expected_freq)


# -----------------------------
# 5. One-way ANOVA
# -----------------------------

group1 = np.array([10, 12, 14, 13, 11])
group2 = np.array([20, 22, 19, 21, 23])
group3 = np.array([30, 28, 32, 31, 29])

anova = stats.f_oneway(group1, group2, group3)

print("\nANOVA:")
print("F-statistic:", anova.statistic)
print("P-value:", anova.pvalue)



Mean: 15.0
Median: 15.0
Population Variance: 11.666666666666666
Population SD: 3.415650255319866
Sample Variance: 14.0
Sample SD: 3.7416573867739413
Q1: 12.5
Q3: 17.5
IQR: 5.0
Skewness: 0.0
Kurtosis: -1.2685714285714282
Z-scores - Population: [-1.46385011 -0.87831007 -0.29277002  0.29277002  0.87831007  1.46385011]
Z-scores - Sample: [-1.33630621 -0.80178373 -0.26726124  0.26726124  0.80178373  1.33630621]

Pearson Correlation:
Correlation coefficient: 0.9901475429766744
P-value: 0.0011722164447168718

Independent t-test:
T-statistic: -7.36381738309777
P-value: 7.888440023273703e-05

Chi-square test:
Chi-square statistic: 15.041666666666668
P-value: 0.00010516355403363114
Degrees of freedom: 1
Expected frequencies:
[[20. 30.]
 [20. 30.]]

ANOVA:
F-statistic: 162.0
P-value: 2.0751610723423055e-09


In [8]:
data = [10, 12, 14, 16, 18]

mean = sum(data) / len(data)

# Population SD
variance = sum((x - mean)**2 for x in data) / len(data)
sd = variance ** 0.5

z_scores = [(x - mean) / sd for x in data]

print(z_scores)

#Note: A Z-score tells you how many standard deviations a value is above or below the mean.

[-1.414213562373095, -0.7071067811865475, 0.0, 0.7071067811865475, 1.414213562373095]


Interpretation -  
-  
If:  
z = 2  
  
the value is 2 standard deviations above the mean.  
  
If:  
z = -1.5  
  
the value is 1.5 standard deviations below the mean.  
  
If:  
z = 0  
  
the value is exactly at the mean.  
  
Important: Z-score itself doesn't have separate "population" and "sample" formulas in the same sense as variance; you use the appropriate mean and SD depending on whether you're standardizing a population or a sample.